File: ClassicalSharp/Rendering/Env/WeatherRenderer.cs
// Copyright 2014-2017 ClassicalSharp | Licensed under BSD-3
using System;
using ClassicalSharp.Events;
using ClassicalSharp.GraphicsAPI;
using ClassicalSharp.Map;
using OpenTK;
using BlockID = System.UInt16;
using BlockRaw = System.Byte;

namespace ClassicalSharp.Renderers {

	public sealed class WeatherRenderer : IGameComponent {
		Game game;
		World map;
		
		int rainTexId, snowTexId;
		int vb;
		public short[] heightmap;
		
		const int extent = 4;
		VertexP3fT2fC4b[] vertices = new VertexP3fT2fC4b[8 * (extent * 2 + 1) * (extent * 2 + 1)];
		double rainAcc;
		Vector3I lastPos = new Vector3I(Int32.MinValue);
		
		public void Init(Game game) {
			this.game = game;
			map = game.World;
			game.Events.TextureChanged += TextureChanged;
			
			ContextRecreated();
			game.Graphics.ContextLost += ContextLost;
			game.Graphics.ContextRecreated += ContextRecreated;
		}
		
		public void Render(double deltaTime) {
			Weather weather = map.Env.Weather;
			if (weather == Weather.Sunny) return;
			if (heightmap == null) InitHeightmap();
			IGraphicsApi gfx = game.Graphics;
			
			gfx.BindTexture(weather == Weather.Rainy ? rainTexId : snowTexId);
			Vector3 camPos = game.CurrentCameraPos;
			Vector3I pos = Vector3I.Floor(camPos);
			bool moved = pos != lastPos;
			lastPos = pos;
			WorldEnv env = game.World.Env;
			
			float speed = (weather == Weather.Rainy ? 1.0f : 0.2f) * env.WeatherSpeed;
			float vOffset = (float)game.accumulator * speed;
			rainAcc += deltaTime;
			bool particles = weather == Weather.Rainy;

			int vCount = 0;
			FastColour col = game.World.Env.Sunlight;
			VertexP3fT2fC4b v = default(VertexP3fT2fC4b);
			
			for (int dx = -extent; dx <= extent; dx++)
				for (int dz = -extent; dz <= extent; dz++)
			{
				int x = pos.X + dx, z = pos.Z + dz;
				float y = RainHeight(x, z);
				float height = Math.Max(game.World.Height, pos.Y + 64) - y;
				if (height <= 0) continue;
				
				if (particles && (rainAcc >= 0.25 || moved))
					game.ParticleManager.AddRainParticle(new Vector3(x, y, z));
				
				float alpha = AlphaAt(dx * dx + dz * dz);
				Utils.Clamp(ref alpha, 0, 0xFF);
				col.A = (byte)alpha;
				
				// NOTE: Making vertex is inlined since this is called millions of times.
				v.Colour = col.Pack();
				float worldV = vOffset + (z & 1) / 2f - (x & 0x0F) / 16f;
				float v1 = y / 6f + worldV, v2 = (y + height) / 6f + worldV;
				float x1 = x,     y1 = y,          z1 = z;
				float x2 = x + 1, y2 = y + height, z2 = z + 1;
				
				v.X = x1; v.Y = y1; v.Z = z1; v.U = 0; v.V = v1; vertices[vCount++] = v;
				v.Y = y2;                    v.V = v2; vertices[vCount++] = v;
				v.X = x2;           v.Z = z2; v.U = 1; 	         vertices[vCount++] = v;
				v.Y = y1;                    v.V = v1; vertices[vCount++] = v;
				
				v.Z = z1;				  	 vertices[vCount++] = v;
				v.Y = y2;                    v.V = v2; vertices[vCount++] = v;
				v.X = x1;           v.Z = z2; v.U = 0;		     vertices[vCount++] = v;
				v.Y = y1;                    v.V = v1; vertices[vCount++] = v;
			}
			if (particles && (rainAcc >= 0.25 || moved)) {
				rainAcc = 0;
			}
			if (vCount == 0) return;
			
			gfx.AlphaTest = false;
			gfx.DepthWrite = false;
			gfx.AlphaArgBlend = true;
			
			gfx.SetBatchFormat(VertexFormat.P3fT2fC4b);
			gfx.UpdateDynamicVb_IndexedTris(vb, vertices, vCount);
			
			gfx.AlphaArgBlend = false;
			gfx.AlphaTest = true;
			gfx.DepthWrite = true;
		}
		
		float AlphaAt(float x) {
			// Wolfram Alpha: fit {0,178},{1,169},{4,147},{9,114},{16,59},{25,9}
			float falloff = 0.05f * x * x - 7 * x;
			return 178 + falloff * game.World.Env.WeatherFade;
		}

		int length, width, maxY, oneY;
		public void Ready(Game game) { }
		public void Reset(Game game) { OnNewMap(game); }
		
		public void OnNewMap(Game game) {
			heightmap = null;
			lastPos = new Vector3I(Int32.MaxValue);
		}
		
		public void OnNewMapLoaded(Game game) {
			length = map.Length;
			width = map.Width;
			maxY = map.Height - 1;
			oneY = length * width;
		}
		
		void TextureChanged(object sender, TextureEventArgs e) {
			if (e.Name == "snow.png")
				game.UpdateTexture(ref snowTexId, e.Name, e.Data, false);
			else if (e.Name == "rain.png")
				game.UpdateTexture(ref rainTexId, e.Name, e.Data, false);
		}
		
		public void Dispose() {
			game.Graphics.DeleteTexture(ref rainTexId);
			game.Graphics.DeleteTexture(ref snowTexId);
			ContextLost();
			
			game.Events.TextureChanged -= TextureChanged;
			game.Graphics.ContextLost -= ContextLost;
			game.Graphics.ContextRecreated -= ContextRecreated;
		}
		
		void InitHeightmap() {
			heightmap = new short[map.Width * map.Length];
			for (int i = 0; i < heightmap.Length; i++) {
				heightmap[i] = short.MaxValue;
			}
		}
		
		float RainHeight(int x, int z) {
			if (x < 0 || z < 0 || x >= width || z >= length) return map.Env.EdgeHeight;
			int index = (x * length) + z;
			int height = heightmap[index];
			int y = height == short.MaxValue ? CalcHeightAt(x, maxY, z, index) : height;
			return y == -1 ? 0 : y + BlockInfo.MaxBB[map.GetBlock(x, y, z)].Y;
		}
		
		int CalcHeightAt(int x, int maxY, int z, int index) {
			int i = (maxY * length + z) * width + x;
			BlockRaw[] blocks = map.blocks;
			
			if (BlockInfo.MaxUsed < 256) {
				for (int y = maxY; y >= 0; y--, i -= oneY) {
					byte draw = BlockInfo.Draw[blocks[i]];
					if (!(draw == DrawType.Gas || draw == DrawType.Sprite)) {
						heightmap[index] = (short)y;
						return y;
					}
				}
			} else {
				BlockRaw[] blocks2 = game.World.blocks2;
				for (int y = maxY; y >= 0; y--, i -= oneY) {
					byte draw = BlockInfo.Draw[blocks[i] | (blocks2[i] << 8)];
					if (!(draw == DrawType.Gas || draw == DrawType.Sprite)) {
						heightmap[index] = (short)y;
						return y;
					}
				}
			}		
			
			heightmap[index] = -1;
			return -1;
		}
		
		internal void OnBlockChanged(int x, int y, int z, BlockID oldBlock, BlockID newBlock) {
			bool didBlock = !(BlockInfo.Draw[oldBlock] == DrawType.Gas || BlockInfo.Draw[oldBlock] == DrawType.Sprite);
			bool nowBlock = !(BlockInfo.Draw[newBlock] == DrawType.Gas || BlockInfo.Draw[newBlock] == DrawType.Sprite);
			if (didBlock == nowBlock) return;
			
			int index = (x * length) + z;
			int height = heightmap[index];
			// Two cases can be skipped here:
			// a) rain height was not calculated to begin with (height is short.MaxValue)
			// b) changed y is below current calculated rain height
			if (y < height) return;
			
			if (nowBlock) {
				// Simple case: Rest of column below is now not visible to rain.
				heightmap[index] = (short)y;
			} else {
				// Part of the column is now visible to rain, we don't know how exactly how high it should be though.
				// However, we know that if the old block was above or equal to rain height, then the new rain height must be <= old block.y
				CalcHeightAt(x, y, z, index);
			}
		}
		
		void ContextLost() { game.Graphics.DeleteVb(ref vb); }
		
		void ContextRecreated() {
			vb = game.Graphics.CreateDynamicVb(VertexFormat.P3fT2fC4b, vertices.Length);
		}
	}
}

File: src/Client/WeatherRenderer.c
#include "WeatherRenderer.h"
#include "Block.h"
#include "ExtMath.h"
#include "Funcs.h"
#include "Event.h"
#include "Game.h"
#include "GraphicsAPI.h"
#include "GraphicsEnums.h"
#include "GraphicsCommon.h"
#include "PackedCol.h"
#include "Platform.h"
#include "Vectors.h"
#include "VertexStructs.h"
#include "World.h"
#include "Particle.h"

GfxResourceID weather_rainTex;
GfxResourceID weather_snowTex;
GfxResourceID weather_vb;

#define weather_extent 4
#define weather_verticesCount 8 * (weather_extent * 2 + 1) * (weather_extent * 2 + 1)

Real64 weather_accumulator;
Vector3I weather_lastPos;

void WeatherRenderer_InitHeightmap(void) {
	Weather_Heightmap = Platform_MemAlloc(World_Width * World_Length * sizeof(Int16));
	if (Weather_Heightmap == NULL) {
		ErrorHandler_Fail("WeatherRenderer - Failed to allocate heightmap");
	}

	Int32 i;
	for (i = 0; i < World_Width * World_Length; i++) {
		Weather_Heightmap[i] = Int16_MaxValue;
	}
}

Int32 WeatherRenderer_CalcHeightAt(Int32 x, Int32 maxY, Int32 z, Int32 index) {
	Int32 mapIndex = (maxY * World_Length + z) * World_Width + x;
	Int32 y = maxY;
	for (y = maxY; y >= 0; y--) {
		UInt8 draw = Block_Draw[World_Blocks[mapIndex]];
		if (!(draw == DRAW_GAS || draw == DRAW_SPRITE)) {
			Weather_Heightmap[index] = (Int16)y;
			return y;
		}
		mapIndex -= World_OneY;
	}
	Weather_Heightmap[index] = -1;
	return -1;
}

Real32 WeatherRenderer_RainHeight(Int32 x, Int32 z) {
	if (x < 0 || z < 0 || x >= World_Width || z >= World_Length) {
		return (Real32)WorldEnv_EdgeHeight;
	}
	Int32 index = (x * World_Length) + z;
	Int32 height = Weather_Heightmap[index];

	Int32 y = height == Int16_MaxValue ? WeatherRenderer_CalcHeightAt(x, World_MaxY, z, index) : height;
	return y == -1 ? 0 : y + Block_MaxBB[World_GetBlock(x, y, z)].Y;
}

void WeatherRenderer_OnBlockChanged(Int32 x, Int32 y, Int32 z, BlockID oldBlock, BlockID newBlock) {
	bool didBlock = !(Block_Draw[oldBlock] == DRAW_GAS || Block_Draw[oldBlock] == DRAW_SPRITE);
	bool nowBlock = !(Block_Draw[newBlock] == DRAW_GAS || Block_Draw[newBlock] == DRAW_SPRITE);
	if (didBlock == nowBlock) return;

	Int32 index = (x * World_Length) + z;
	Int32 height = Weather_Heightmap[index];
	/* Two cases can be skipped here: */
	/* a) rain height was not calculated to begin with (height is short.MaxValue) */
	/* b) changed y is below current calculated rain height */
	if (y < height) return;

	if (nowBlock) {
		/* Simple case: Rest of column below is now not visible to rain. */
		Weather_Heightmap[index] = (Int16)y;
	} else {
		/* Part of the column is now visible to rain, we don't know how exactly how high it should be though. */
		/* However, we know that if the old block was above or equal to rain height, then the new rain height must be <= old block.y */
		WeatherRenderer_CalcHeightAt(x, y, z, index);
	}
}

void WeatherRenderer_ContextLost(void* obj) {
	Gfx_DeleteVb(&weather_vb);
}

void WeatherRenderer_ContextRecreated(void* obj) {
	weather_vb = Gfx_CreateDynamicVb(VERTEX_FORMAT_P3FT2FC4B, weather_verticesCount);
}

Real32 WeatherRenderer_AlphaAt(Real32 x) {
	/* Wolfram Alpha: fit {0,178},{1,169},{4,147},{9,114},{16,59},{25,9} */
	Real32 falloff = 0.05f * x * x - 7 * x;
	return 178 + falloff * WorldEnv_WeatherFade;
}

void WeatherRenderer_Render(Real64 deltaTime) {
	Int32 weather = WorldEnv_Weather;
	if (weather == WEATHER_SUNNY) return;
	if (Weather_Heightmap == NULL) WeatherRenderer_InitHeightmap();

	Gfx_BindTexture(weather == WEATHER_RAINY ? weather_rainTex : weather_snowTex);
	Vector3 camPos = Game_CurrentCameraPos;
	Vector3I pos;
	Vector3I_Floor(&pos, &camPos);
	bool moved = Vector3I_NotEquals(&pos, &weather_lastPos);
	weather_lastPos = pos;

	/* Rain should extend up by 64 blocks, or to the top of the world. */
	pos.Y += 64;
	pos.Y = max(World_Height, pos.Y);

	Real32 speed = (weather == WEATHER_RAINY ? 1.0f : 0.2f) * WorldEnv_WeatherSpeed;
	Real32 vOffset = (Real32)Game_Accumulator * speed;
	weather_accumulator += deltaTime;
	bool particles = weather == WEATHER_RAINY;

	PackedCol col = WorldEnv_SunCol;
	VertexP3fT2fC4b v;
	VertexP3fT2fC4b vertices[weather_verticesCount];
	VertexP3fT2fC4b* ptr = vertices;

	Int32 dx, dz;
	for (dx = -weather_extent; dx <= weather_extent; dx++) {
		for (dz = -weather_extent; dz <= weather_extent; dz++) {
			Int32 x = pos.X + dx, z = pos.Z + dz;
			Real32 y = WeatherRenderer_RainHeight(x, z);
			Real32 height = pos.Y - y;
			if (height <= 0) continue;

			if (particles && (weather_accumulator >= 0.25 || moved)) {
				Vector3 particlePos = Vector3_Create3((Real32)x, y, (Real32)z);
				Particles_RainSnowEffect(particlePos);
			}

			Real32 dist = (Real32)dx * (Real32)dx + (Real32)dz * (Real32)dz;
			Real32 alpha = WeatherRenderer_AlphaAt(dist);
			/* Clamp between 0 and 255 */
			alpha = alpha < 0.0f ? 0.0f : alpha;
			alpha = alpha > 255.0f ? 255.0f : alpha;
			col.A = (UInt8)alpha;

			/* NOTE: Making vertex is inlined since this is called millions of times. */
			v.Col = col;
			Real32 worldV = vOffset + (z & 1) / 2.0f - (x & 0x0F) / 16.0f;
			Real32 v1 = y / 6.0f + worldV, v2 = (y + height) / 6.0f + worldV;
			Real32 x1 = (Real32)x,       y1 = (Real32)y,            z1 = (Real32)z;
			Real32 x2 = (Real32)(x + 1), y2 = (Real32)(y + height), z2 = (Real32)(z + 1);

			v.X = x1; v.Y = y1; v.Z = z1; v.U = 0.0f; v.V = v1; *ptr = v; ptr++;
			          v.Y = y2;                      v.V = v2;  *ptr = v; ptr++;
			v.X = x2;           v.Z = z2; v.U = 1.0f; 	        *ptr = v; ptr++;
			          v.Y = y1;                      v.V = v1;  *ptr = v; ptr++;

			                    v.Z = z1;					    *ptr = v; ptr++;
			          v.Y = y2;                       v.V = v2; *ptr = v; ptr++;
			v.X = x1;           v.Z = z2; v.U = 0.0f;		    *ptr = v; ptr++;
			          v.Y = y1;                       v.V = v1; *ptr = v; ptr++;
		}
	}

	if (particles && (weather_accumulator >= 0.25f || moved)) {
		weather_accumulator = 0;
	}
	if (ptr == vertices) return;

	Gfx_SetAlphaTest(false);
	Gfx_SetDepthWrite(false);
	Gfx_SetAlphaArgBlend(true);

	Gfx_SetBatchFormat(VERTEX_FORMAT_P3FT2FC4B);
	UInt32 vCount = (UInt32)(ptr - vertices) / VertexP3fT2fC4b_Size;
	GfxCommon_UpdateDynamicVb_IndexedTris(weather_vb, vertices, vCount);

	Gfx_SetAlphaArgBlend(false);
	Gfx_SetDepthWrite(false);
	Gfx_SetAlphaTest(false);
}

void WeatherRenderer_FileChanged(void* obj, Stream* stream) {
	String snow = String_FromConst("snow.png");
	String rain = String_FromConst("rain.png");

	if (String_Equals(&stream->Name, &snow)) {
		Game_UpdateTexture(&weather_snowTex, stream, false);
	} else if (String_Equals(&stream->Name, &rain)) {
		Game_UpdateTexture(&weather_rainTex, stream, false);
	}
}

void WeatherRenderer_Init(void) {
	weather_lastPos = Vector3I_Create1(Int32_MaxValue);
	WeatherRenderer_ContextRecreated(NULL);

	Event_RegisterStream(&TextureEvents_FileChanged, NULL, &WeatherRenderer_FileChanged);
	Event_RegisterVoid(&GfxEvents_ContextLost,       NULL, &WeatherRenderer_ContextLost);
	Event_RegisterVoid(&GfxEvents_ContextRecreated,  NULL, &WeatherRenderer_ContextRecreated);
}

void WeatherRenderer_Reset(void) {
	if (Weather_Heightmap != NULL) Platform_MemFree(Weather_Heightmap);
	Weather_Heightmap = NULL;
	weather_lastPos = Vector3I_Create1(Int32_MaxValue);
}

void WeatherRenderer_Free(void) {
	Gfx_DeleteTexture(&weather_rainTex);
	Gfx_DeleteTexture(&weather_snowTex);
	WeatherRenderer_ContextLost(NULL);
	WeatherRenderer_Reset();

	Event_UnregisterStream(&TextureEvents_FileChanged, NULL, &WeatherRenderer_FileChanged);
	Event_UnregisterVoid(&GfxEvents_ContextLost,       NULL, &WeatherRenderer_ContextLost);
	Event_UnregisterVoid(&GfxEvents_ContextRecreated,  NULL, &WeatherRenderer_ContextRecreated);
}

IGameComponent WeatherRenderer_MakeGameComponent(void) {
	IGameComponent comp = IGameComponent_MakeEmpty();
	comp.Init = WeatherRenderer_Init;
	comp.Free = WeatherRenderer_Free;
	comp.OnNewMap = WeatherRenderer_Reset;
	comp.Reset = WeatherRenderer_Reset;
	return comp;
}